pkg/kubelet/util/pod_startup_latency_tracker_test.go

```diff
@@ -275,6 +275,72 @@ kubelet_pod_start_sli_duration_seconds_count 1
+
+		wants := `
+# HELP kubelet_first_network_pod_start_sli_duration_seconds [ALPHA] Duration in seconds to start the first network pod, excluding time to pull images and run init containers, measured from pod creation timestamp to when all its containers are reported as started and observed via watch
+# TYPE kubelet_first_network_pod_start_sli_duration_seconds gauge
+kubelet_first_network_pod_start_sli_duration_seconds 30
+# HELP kubelet_first_network_pod_start_total_duration_seconds [ALPHA] Duration in seconds to start the first network pod since creation, including time to pull images and run init containers, measured from pod creation timestamp to when all its containers are reported as started and observed via watch
+# TYPE kubelet_first_network_pod_start_total_duration_seconds gauge
+kubelet_first_network_pod_start_total_duration_seconds 30
+`
+
+		fakeClock := testingclock.NewFakeClock(frozenTime)
+
+		metrics.Register()
+
+		tracker := &basicPodStartupLatencyTracker{
+			pods:  map[types.UID]*perPodState{},
+			clock: fakeClock,
+		}
+
+		// hostNetwork pods should not be tracked
+		hostNetworkPodInitializing := buildInitializingPod()
+		hostNetworkPodInitializing.UID = "11111-22222"
+		hostNetworkPodInitializing.Spec.HostNetwork = true
+		tracker.ObservedPodOnWatch(hostNetworkPodInitializing, frozenTime)
+
+		hostNetworkPodStarted := buildRunningPod()
+		hostNetworkPodStarted.UID = "11111-22222"
+		hostNetworkPodStarted.Spec.HostNetwork = true
+		tracker.RecordStatusUpdated(hostNetworkPodStarted)
+
+		// track only the first pod with network
+		podInitializing := buildInitializingPod()
+		tracker.ObservedPodOnWatch(podInitializing, frozenTime)
+
+		// pod started
+		podStarted := buildRunningPod()
+		tracker.RecordStatusUpdated(podStarted)
+
+		// at 30s observe the same pod on watch
+		tracker.ObservedPodOnWatch(podStarted, frozenTime.Add(time.Second*30))
+
+		if err := testutil.GatherAndCompare(metrics.GetGather(), strings.NewReader(wants), "kubelet_first_network_pod_start_sli_duration_seconds", "kubelet_first_network_pod_start_total_duration_seconds"); err != nil {
+			t.Fatal(err)
+		}
+
+		// any new pod observations should not impact the metrics, as the pod should be recorder only once
+		tracker.ObservedPodOnWatch(podStarted, frozenTime.Add(time.Second*150))
+		tracker.ObservedPodOnWatch(podStarted, frozenTime.Add(time.Second*200))
+		tracker.ObservedPodOnWatch(podStarted, frozenTime.Add(time.Second*250))
+
+		if err := testutil.GatherAndCompare(metrics.GetGather(), strings.NewReader(wants), "kubelet_first_network_pod_start_sli_duration_seconds", "kubelet_first_network_pod_start_total_duration_seconds"); err != nil {
+			t.Fatal(err)
+		}
+
+		// cleanup
+		tracker.DeletePodStartupState(hostNetworkPodStarted.UID)
+		tracker.DeletePodStartupState(podStarted.UID)
+
+		assert.Empty(t, tracker.pods)
+		metrics.PodStartSLIDuration.Reset()
```

Comment:
Why are you resetting this metric, you're not even testing this one... shouldn't you be setting the gauge values of the tested metrics to 0?

Reply:
the metrics has to be reseted always or when you do some test with `-count N` or with other test with metrics in parallel they fail, since the metrics registry is global, see https://github.com/kubernetes/kubernetes/pull/100218